activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1915,7 +1915,14 @@ def build_with_expression_from_value(value)
         when Arel::Nodes::SqlLiteral then Arel::Nodes::Grouping.new(value)
         when ActiveRecord::Relation then value.arel
         when Arel::SelectManager then value
-        when Array then value.map { |q| build_with_expression_from_value(q) }.reduce { |result, value| result.union(:all, value) }
+        when Array
+          parts = value.map { |q| build_with_expression_from_value(q) }
+          parts.reduce do |result, value|
+            Arel::Nodes::UnionAll.new(
+              result.try(:ast) || result,
```

Comment:
Not a big fan of using `.try ||` here. It's often indicative of unclear APIs.

Reply:
@fatkodima I pushed 9d9414b35e592362eb8429308d6103e586ecf23c, let me know what you think.